### PR TITLE
use containerd/containerd/api module

### PR DESCRIPTION
**Pull request by akhilerm**
- use containerd/containerd/api module instead of having a copy in third_party
- use containerd/errdefs module instead of keeping a local copy

### CODE REVIEW

**Comment by bobbypage** [comment]
Very exciting to see this happen!

**Comment by akhilerm** [comment]
@bobbypage Can you approve the github action workflows also?

File: build/integration-in-docker.sh
Comment (akhilerm) [inline build/integration-in-docker.sh:63]: This is being fixed in #3527 to switch to bookworm

**Comment by akhilerm** [comment]
/hold
will merge once containerd/containerd/api v1.8.0 is released

**Comment by akhilerm** [comment]
@bobbypage Can you approve the workflow. I have rebased against the latest master branch

**Comment by iwankgb** [comment]
@akhilerm, you have conflicts against master.

**Comment by akhilerm** [comment]
rebased against master and pushed the changes.

**Comment by akhilerm** [comment]
The `pull-cadvisor-e2e ` job is failing due to env setup error

/retest pull-cadvisor-e2e

File: container/containerd/containers/containers.go
Comment (iwankgb) [inline container/containerd/containers/containers.go:36]: Is this alias necessary?

File: container/containerd/handler_test.go
Comment (iwankgb) [inline container/containerd/handler_test.go:75]: Can these two functions be unexported, please?

**Review by iwankgb — COMMENTED** [review]
Just two minor comments :)

File: cmd/go.mod
Comment (iwankgb) [inline cmd/go.mod:47]: This PR shall not be merged with RC version.

**Review by iwankgb — CHANGES_REQUESTED** [review]

File: container/containerd/handler_test.go
Comment (akhilerm) [reply container/containerd/handler_test.go:75]: There is a PR in https://github.com/containerd/typeurl/pull/45, so that these funcs can be vendored and used instead of copying it over here.

File: cmd/go.mod
Comment (akhilerm) [reply cmd/go.mod:47]: This PR was raised so as to test the containerd/api module compatiblity with cadvisor.  I am updating as each rc version releases so that we know there are no breaking changes.  

PR needs to be merged only after containerd/api v1.8.0 is released. 

More discussion on the same is happening on this thread https://kubernetes.slack.com/archives/CHGFYJVAN/p1714733633634819

File: container/containerd/containers/containers.go
Comment (akhilerm) [reply container/containerd/containers/containers.go]: FIxed.

File: cmd/go.mod
Comment (thaJeztah) [reply cmd/go.mod:47]: Note that API v1.7.19 has been tagged; the v1.8.0-rc is the API version from the main branch (containerd v2.0-dev), so depending on what's needed, maybe the v1.7.19 version is an option?

Comment (akhilerm) [reply cmd/go.mod:47]: Yes. Thats a better option. Will update to api/v1.7.19

Comment (akhilerm) [reply cmd/go.mod:47]: Done.

**Comment by akhilerm** [comment]
/hold

for https://github.com/containerd/typeurl/pull/45 to merge

File: cmd/go.mod
Comment (thaJeztah) [reply cmd/go.mod:47]: Looks like the "indirect" here stuck on 1.8; could be either go modules's tendency to not downgrade once it's there, or could there be something else in the dependency tree pushing it up? 🤔 

(Sorry, maybe you already noticed that, but just in case 🙈 🫶 )

Comment (akhilerm) [reply cmd/go.mod]: My bad. Hadnt checked the new diff before pushing. Thought tidy will downgrade it automatically. It was because of this

> could be either go modules's tendency to not downgrade once it's there

running just a `go mod tidy` will not downgrade, had to remove and run the tidy again so as to bring in 1.7.19 as indirect dependency.

Comment (thaJeztah) [reply cmd/go.mod]: Yes, that fooled me many times; if a dependency version (based on the dependency tree "minimum version selection")  either becomes lower, or even is no longer a dependency, then go modules thinks it was a manual override / upgrade.

That sometimes makes sense (if you indeed updated the version because you wanted a newer version) but sometimes you don't, and just remove all indirects to force it re-resolving the minimum versions can help find those.

**Review by iwankgb — APPROVED** [review]

**Comment by thaJeztah** [comment]
Hmm looks like some images used in tests are using a deprecated format, or is it detecting it wrong?

```
--- PASS: TestDockerContainerSpec (0.40s)
=== RUN   TestDockerContainerCpuStats
    framework.go:337: Failed to run "sudo" [docker run -d registry.k8s.io/busybox ping www.google.com] in "localhost" with error: "exit status 125". Stdout: "", Stderr: Unable to find image 'registry.k8s.io/busybox:latest' locally
        latest: Pulling from busybox
        docker: [DEPRECATION NOTICE] Docker Image Format v1 and Docker Image manifest version 2, schema 1 support is disabled by default and will be removed in an upcoming release. Suggest the author of registry.k8s.io/busybox:latest to upgrade the image to the OCI Format or Docker Image manifest v2, schema 2. More information at [https://docs.docker.com/go/deprecated-image-specs/.](https://docs.docker.com/go/deprecated-image-specs/)
        See 'docker run --help'.

**Comment by iwankgb** [comment]
@thaJeztah, it's fixed on `master` :)

**Comment by akhilerm** [comment]
rebased and pushed.

**Comment by akhilerm** [comment]
@iwankgb Can you approve the workflows once more. 

Once we have a solution for https://github.com/kubernetes-sigs/prow/issues/194, one time approval should do the job.

**Comment by iwankgb** [comment]
@akhilerm, you have some conflicts in `go.mod` :(

**Review by iwankgb — APPROVED** [review]

**Pull request merged**
